kube-client/src/api/portforward.rs

```diff
@@ -0,0 +1,335 @@
+        }
+
+        let state = Arc::new(Mutex::new(PortforwarderState {
+            waker: None,
+            result: None,
+        }));
+        let shared_state = state.clone();
+        let port_nums = port_nums.to_owned();
+        tokio::spawn(async move {
+            let result = start_message_loop(stream, port_nums, duplexes, errors).await;
+
+            let mut shared = shared_state.lock().unwrap();
+            shared.result = Some(result);
+            if let Some(waker) = shared.waker.take() {
+                waker.wake()
+            }
+        });
+        Portforwarder { ports, state }
+    }
+
+    /// Get streams for forwarded ports.
+    pub fn ports(&mut self) -> &mut [Port] {
+        self.ports.as_mut_slice()
+    }
+}
+
+impl Future for Portforwarder {
+    type Output = Result<(), Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut state = self.state.lock().unwrap();
+        if let Some(result) = state.result.take() {
+            return Poll::Ready(result);
+        }
+
+        if let Some(waker) = &state.waker {
+            if waker.will_wake(cx.waker()) {
+                return Poll::Pending;
+            }
+        }
+
+        state.waker = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+pub struct Port {
+    // Data pipe.
+    stream: Option<DuplexStream>,
+    // Error channel.
+    error: Option<ErrorReceiver>,
+}
+
+impl Port {
+    pub(crate) fn new(stream: DuplexStream, error: ErrorReceiver) -> Self {
+        Port {
+            stream: Some(stream),
+            error: Some(error),
+        }
+    }
+
+    /// Data pipe for sending to and receiving from the forwarded port.
+    ///
+    /// This returns a `Some` on the first call, then a `None` on every subsequent call
+    pub fn stream(&mut self) -> Option<impl AsyncRead + AsyncWrite + Unpin> {
+        self.stream.take()
+    }
+
+    /// Future that resolves with any error message or when the error sender is dropped.
+    /// When the future resolves, the port should be considered no longer usable.
+    ///
+    /// This returns a `Some` on the first call, then a `None` on every subsequent call
+    pub fn error(&mut self) -> Option<impl Future<Output = Option<String>>> {
+        // Ignore Cancellation error.
+        self.error.take().map(|recv| recv.map(|res| res.ok()))
+    }
+}
+
+async fn start_message_loop<S>(
+    stream: WebSocketStream<S>,
+    ports: Vec<u16>,
+    duplexes: Vec<DuplexStream>,
+    error_senders: Vec<Option<ErrorSender>>,
+) -> Result<(), Error>
+where
+    S: AsyncRead + AsyncWrite + Unpin + Sized + Send + 'static,
+{
+    let mut writers = Vec::new();
+    // Loops to run concurrently.
+    // We can spawn tasks to run `to_pod_loop` in parallel and flatten the errors, but the other 2 loops
+    // are over a single WebSocket connection and cannot process each port in parallel.
+    let mut loops = Vec::with_capacity(ports.len() + 2);
```

Comment:
Changed not to spawn tasks because other loops can't run in parallel, and it's easier to manage.

Reply:
Yeah, it's a lot nicer.